python_modules/libraries/dagster-docker/dagster_docker/external_resource.py

```diff
@@ -1,103 +1,127 @@
-                    print(line)
-
-                result = container.wait()
-                return result["StatusCode"]
-            finally:
-                container.stop()
+            result = container.wait()
+            if result["StatusCode"] != 0:
+                raise DagsterExternalError(f"Container exited with non-zero status code: {result}")
+        finally:
+            container.stop()
+
+    # ########################
+    # ##### IO CONTEXT MANAGERS
+    # ########################
+
+    def _input_context_manager(
+        self, tempdir: str, sockaddr: SocketAddress
+    ) -> ContextManager[DockerTaskIOParams]:
+        if self._input_mode == ExternalExecutionIOMode.file:
+            return self._file_input(tempdir)
+        elif self._input_mode == ExternalExecutionIOMode.socket:
+            return self._socket_input_or_output(sockaddr)
+        else:
+            check.failed(f"Unsupported input mode: {self._input_mode}")
+
+    @contextmanager
+    def _file_input(self, tempdir: str) -> Iterator[DockerTaskIOParams]:
+        path = self._prepare_io_path(self._input_path, "input", tempdir)
+        env = {DAGSTER_EXTERNAL_ENV_KEYS["input"]: path}
+        try:
+            self._write_input(path)
+            path_dir = os.path.dirname(path)
+            volume_mounts = {path_dir: {"bind": path_dir, "mode": "rw"}}
+            yield DockerTaskIOParams(env=env, volume_mounts=volume_mounts)
+        finally:
+            if os.path.exists(path):
+                os.remove(path)
+
+    def _output_context_manager(
+        self, tempdir: str, sockaddr: SocketAddress
+    ) -> ContextManager[DockerTaskIOParams]:
+        if self._output_mode == ExternalExecutionIOMode.file:
+            return self._file_output(tempdir)
+        elif self._output_mode == ExternalExecutionIOMode.socket:
+            return self._socket_input_or_output(sockaddr)
+        else:
+            check.failed(f"Unsupported output mode: {self._output_mode}")
+
+    @contextmanager
+    def _file_output(self, tempdir: str) -> Iterator[DockerTaskIOParams]:
+        path = self._prepare_io_path(self._output_path, "output", tempdir)
+        env = {DAGSTER_EXTERNAL_ENV_KEYS["output"]: path}
+        output_file_dir = os.path.dirname(path)
+        volume_mounts = {output_file_dir: {"bind": output_file_dir, "mode": "rw"}}
+        try:
+            yield DockerTaskIOParams(env=env, volume_mounts=volume_mounts)
+            self._read_output(path)
+        finally:
+            os.remove(path)
+
+    @contextmanager
+    def _socket_input_or_output(self, sockaddr: SocketAddress) -> Iterator[DockerTaskIOParams]:
+        _, port = sockaddr
+        ipc_ports = {port: port}
+        env = {
+            DAGSTER_EXTERNAL_ENV_KEYS["host"]: "host.docker.internal",
+            DAGSTER_EXTERNAL_ENV_KEYS["port"]: port,
+        }
+        yield DockerTaskIOParams(env=env, ports=ipc_ports)
 
 
 class DockerExecutionResource(ExternalExecutionResource):
```

Comment:
any thoughts on how to improve this layer ? Should `ExternalExecutionIOMode` be shared or should we have separate mode enums? Can you compose config nicely along `ConfigurableResource`  inheritance or should we change tactics here

Reply:
>Should ExternalExecutionIOMode be shared or should we have separate mode enums?

It should be shared because it is used by both the orchestration and transform process-- i.e. `dagster-externals` uses it and that is shared between all orch-side adapters (what are we calling these things?)

>any thoughts on how to improve this layer? ... Can you compose config nicely along ConfigurableResource inheritance or should we change tactics here

Not clear on exactly what you mean by "compose config nicely along ConfigurableResource inheritance". If the question is just whether you can add additional fields with `ConfigurableResource` inheritance, the answer is yes.

Reply:
what I was thinking about was that the valid input/output modes differ between docker and subprocess, and ideally thats reflected in the config. I guess one answer is to sort of mirror the `TaskIOParams` tree with these resources and only have the modes defined on the subprocess/docker resources

Reply:
>what I was thinking about was that the valid input/output modes differ between docker and subprocess, and ideally thats reflected in the config.

Yeah this is a good point.

I think one solution is to type the resources with `Literal` instead of the Enum class and just use the enum class inside our internal machinery:

```
class DockerExecutionResource(ExternalExecutionResource):
    input_mode: Literal["file", "socket"]  # overrides input mode from above
```

That currently doesn't work but I think it might be a minor addition to the config machinery to make it work, since that machinery does accept enums and this is effectively an inline enum.